### DISCUSSION TITLE
Tropico - hardware rendering mode yields corrupted graphics

### DISCUSSION THREAD

**Discussion by Ligushka** [elishacloud/dxwrapper#109] [Q&A] [answered]
The 2001 simulation game Tropico works fine in modern Windows versions using software rendering. However, when hardware rendering is attempted, graphics are mostly corrupted, as seen in the screenshot.

Is there a way to use dxwrapper to correct the problem?

![tropico-hardware](https://user-images.githubusercontent.com/13472038/103966825-a6941480-512e-11eb-88c8-c527c713b7e9.png)

**Comment by elishacloud** [comment]
This issue is almost certainly related to [pitch/stride](https://github.com/elishacloud/dxwrapper/blob/802af256d38994835243b47958451e626e28e789/ddraw/IDirectDrawTypes.cpp#L790-L793).  I have seen this issue before.  What version of DirectX does Tropica use?

**Reply by Ligushka** [reply]
Sorry for the delayed reply, Github did not send me an email notification.

My understanding is that Tropico uses Direct3D 7.

**Reply by Ligushka** [reply]
I just checked the Tropico demo and it exhibits similar behavior when using Hardware rendering. The demo can be downloaded from CNET: https://download.cnet.com/Tropico-demo/3055-2119_4-10059685.html

**Reply by elishacloud** [reply]
Thanks for the link to the demo!

**Comment by elishacloud** [comment, marked as answer]
The issue is that the game completely ignores the pitch/stride sent by DirectDraw.  In newer versions of Windows the surfaces are byte aligned for performance reasons and so the pitch may be larger than the width.  In this case some of the game's video data is written to the unused part of the surface.

I added a quick fix for this by emulating the lock function and feeding the game a pitch that it is expecting.

Here is the updated build:  [dxwrapper.zip](https://github.com/elishacloud/dxwrapper/files/6269166/dxwrapper.zip)

**Reply by Ligushka** [reply]
Superb, this seems to work perfectly in the full version (I have the GOG version)! Great job.

The only downside to using the hardware rendering mode is the lack of water animations now, which the software renderer has. I wouldn't think that would be something that is correctable by dxwrapper, though. My recollection is that this game has always behaved this way - hardware rendering looks nicer (e.g. beachline meeting the water is much smoother), but the water animations are inexplicably unsupported.

**Reply by msbr0815** [reply]
Hi, thanks a lot for providing the quick fix. I hope it's ok if I'm contributing a few thoughts to this old discussion, but my observations may be interesting to some people.

There seem to be several versions of the game, some of which may not load this wrapper (or any other for that matter) and thus may not correct the graphical errors described by Ligushka further above. My copy, which I bought on Green Man Gaming in 2014, appears to be such a case: It includes some sort of securom copyright protection and needs to be activated before playing on a new computer. Consequently, the size of my Tropico.exe is a whopping 10.6 MB, whereas others are closer to 2 MB, such as the v1.53 patch of the Mucho Macho version I found through PCgamingwiki. I would very much like to use the hardware acceleration, since the game is essentially an unplayable slide show in software rendering mode.

I understand that the dll files in the same directory as the exe file are supposed to be loaded automatically when starting the game, but for some reason, my particular version of Tropico Reloaded does not seem to do so (dxwrapper.dll, dxwrapper.ini and d3draw.dll are in the same directory as tropico.exe):
- None of the recommended settings in dxwrapper.ini removed the pitch/stride glitches
- Without compatibility settings, he game starts in windowed mode and I cannot seem to force it to fullscreen using the [FullScreen] setting
- No log file is created, although I left `DisableLogging = 0`

When replacing my executable with the one from the 1.53 patch mentioned above, dxwrapper does create a log file, but the game won't start, since I don't have the CD. Am I correct in assuming that the behaviour I observe with my the version I own should be a consequence of the copy protection, i. e. of some non-tamper mechanism? Would there be another possibility to get the game to run smoothly or should I grasp the nettle and wait for a good deal on gog?

Just another small observation which may be unrelated: I added the line `RunProcess = powershell.exe "pause"` to dxwrapper.ini and would have expected to see a powershell window asking me to press any key to continue upon quitting or minimising the game. This windows doesn't appear using either of the tested versions. Did I make a mistake somewhere?

EDIT: small corrections

**Reply by elishacloud** [reply]
I guess the game is loading ddraw.dll from the system folder.  Try renaming the ddraw.dll stub file to something else.  Keep dxwrapper.dll and dxwrapper.ini the same.

Here are some possible options for renaming ddraw.dll:  `dsound.dll`, `dinput.dll`, `winmm.dll`, `dciman32.dll`

**Reply by msbr0815** [reply]
Wow, thanks a lot for your quick reply – it is working when I rename the file to `dsound.dll` (but none of the others). The Windows XP compatibility mode is required, but I expected that. The system indeed loads `ddraw.dll` from the system folder according to the log file. Again, thanks a lot, now I can finally play this game on my current setup! :)

Out of interest, would it somehow be possible to rename `ddraw.dll` to something else and load it through the ini file instead of renaming it to `dsound.dll`? I gave it a quick try, but couldn't get it to work. I may not have used the proper syntax, though, and since your suggestion is working, I didn't want to spend too much time on it.

**Reply by elishacloud** [reply]
There are a number of other file names you can use to rename ddraw.dll to.  I just picked those ones because they are the best chance of working.  Is there something wrong with using dsound.dll?

BTW: if you want to use a different dsound.dll file you can use the dsound.dll ini file to redirect it to any other file.  See here:  https://github.com/elishacloud/dxwrapper/wiki/Configuration-(stub)

**Reply by msbr0815** [reply]
Nothing wrong at all, I was just curious to see what was possible, e. g. in case I'd run into similar problems with other games in the future.

**Reply by elishacloud** [reply]
The possible file names are mentioned in the link I gave above:  https://github.com/elishacloud/dxwrapper/wiki/Configuration-(stub)

**Comment by Ligushka** [comment]
Somewhat related, is there a way to use dxwrapper to have the 1280x1024 resolution "boxed in" at the native res of 1920x1080 with black filling in the rest of the screen? Enabling dd7to9 and then these settings yields the correct look, but also results in the mouse cursor having an "offset" / binding to what seems to be a 640x480 box in the upper left corner of the main menu, plus Tropico believes hardware mode is not usable when using dd7to9 without these settings defined.

DdrawUseNativeResolution = 1
DdrawIntegerScalingClamp = 1
DdrawMaintainAspectRatio = 1

I can adjust my monitor's settings to display 5:4 aspect ratio instead as a workaround if there's no way around the resolution limitation.

Thank you again for your time and effort. I find dxwrapper indispensable for playing Star Trek: Armada (1) as well.

**Reply by elishacloud** [reply]
> Somewhat related, is there a way to use dxwrapper to have the 1280x1024 resolution "boxed in" at the native res of 1920x1080 with black filling in the rest of the screen?

I spent sometime looking into this.  The game handles the mouse using hardware rather than using DirectX.  This means if you stretch the game graphics it will not line up with the mouse.  I don't have code in dxwrapper to handle games that use mouse hardware rather than DirectX.  I might look into this later, but it is not really the purpose of dxwrapper.

> plus Tropico believes hardware mode is not usable when using dd7to9 without these settings defined.

This is happening because when using hardware mode it uses Direct3D APIs.  I have not yet added support for Direct3D APIs when using dd7to9.  I started that code earlier this year, but it will take a good amount of time before it is all completed.

**Reply by Ligushka** [reply]
Thanks for taking the time to check that out. I understand, I'll adjust aspect ratio on my monitor instead.

I've spent about 3 hours playing Tropico using the hardware rendering mode with dxwrapper and it's working well and stable. Thank you again.
